### PR TITLE
DiskWeb fix

### LINKED ISSUE #57188: DiskWeb Logical error: 'std::exception. Code: 1001, type: std::out_of_range, e.what() = map::at:  key not found (version 23.10.1.1), Stack trace:
Triggered by `test_disk_over_web_server` test with debug build
>           node2.query("SELECT count() FROM system.remote_data_paths")

```
E           helpers.client.QueryRuntimeException: Client failed! Return code: 32, stderr: [node2] 2023.11.24 08:50:11.344503 [ 12 ] {b35e75e9-c0f1-4eb2-9a12-d4864a699e42} <Fatal> : Logical error: 'std::exception. Code: 1001, type: std::out_of_range, e.what() = map::at:  key not found (version 23.10.1.1), Stack trace:
E
E           0. /home/ubuntu/clickhouse/contrib/llvm-project/libcxx/include/exception:134: std::exception::capture() @ 0x000000000b314402 in /usr/bin/clickhouse
E           1. /home/ubuntu/clickhouse/contrib/llvm-project/libcxx/include/exception:112: std::exception::exception[abi:v15000]() @ 0x000000000b3143cd in /usr/bin/clickhouse
E           2. /home/ubuntu/clickhouse/contrib/llvm-project/libcxx/src/support/runtime/stdexcept_default.ipp:24: std::logic_error::logic_error(char const*) @ 0x0000000029047afd in /usr/bin/clickhouse
E           3. /home/ubuntu/clickhouse/contrib/llvm-project/libcxx/include/stdexcept:166: std::out_of_range::out_of_range[abi:v15000](char const*) @ 0x000000000bb2b061 in /usr/bin/clickhouse
E           4. /home/ubuntu/clickhouse/contrib/llvm-project/libcxx/include/stdexcept:268: std::__throw_out_of_range[abi:v15000](char const*) @ 0x000000000bb2afe9 in /usr/bin/clickhouse
E           5. /home/ubuntu/clickhouse/contrib/llvm-project/libcxx/include/map:1655: std::map<String, DB::WebObjectStorage::FileData, std::less<String>, std::allocator<std::pair<String const, DB::WebObjectStorage::FileData>>>::at(String const&) @ 0x000000001be999ff in /usr/bin/clickhouse
E           6. /home/ubuntu/clickhouse/src/Disks/ObjectStorages/Web/MetadataStorageFromStaticFilesWebServer.cpp:63: DB::MetadataStorageFromStaticFilesWebServer::isFile(String const&) const @ 0x000000001be9842b in /usr/bin/clickhouse
E           7. /home/ubuntu/clickhouse/src/Disks/ObjectStorages/DiskObjectStorage.cpp:79: DB::DiskObjectStorage::getRemotePathsRecursive(String const&, std::vector<DB::IDisk::LocalPathWithObjectStoragePaths, std::allocator<DB::IDisk::LocalPathWithObjectStoragePaths>>&) @ 0x000000001be2f04d in /usr/bin/clickhouse
E           8. /home/ubuntu/clickhouse/src/Storages/System/StorageSystemRemoteDataPaths.cpp:60: DB::StorageSystemRemoteDataPaths::read(std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::StorageSnapshot> const&, DB::SelectQueryInfo&, std::shared_ptr<DB::Context const>, DB::QueryProcessingStage::Enum, unsigned long, unsigned long) @ 0x000000001ade4370 in /usr/bin/clickhouse
E           9. /home/ubuntu/clickhouse/src/Storages/IStorage.cpp:144: DB::IStorage::read(DB::QueryPlan&, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::StorageSnapshot> const&, DB::SelectQueryInfo&, std::shared_ptr<DB::Context const>, DB::QueryProcessingStage::Enum, unsigned long, unsigned long) @ 0x000000001d98ab61 in /usr/bin/clickhouse
E           10. /home/ubuntu/clickhouse/src/Interpreters/InterpreterSelectQuery.cpp:2486: DB::InterpreterSelectQuery::executeFetchColumns(DB::QueryProcessingStage::Enum, DB::QueryPlan&) @ 0x000000001ce0d2cf in /usr/bin/clickhouse
E           11. /home/ubuntu/clickhouse/src/Interpreters/InterpreterSelectQuery.cpp:0: DB::InterpreterSelectQuery::executeImpl(DB::QueryPlan&, std::optional<DB::Pipe>) @ 0x000000001ce052aa in /usr/bin/clickhouse
E           12. /home/ubuntu/clickhouse/src/Interpreters/InterpreterSelectQuery.cpp:868: DB::InterpreterSelectQuery::buildQueryPlan(DB::QueryPlan&) @ 0x000000001ce043fd in /usr/bin/clickhouse
E           13. /home/ubuntu/clickhouse/src/Interpreters/InterpreterSelectWithUnionQuery.cpp:303: DB::InterpreterSelectWithUnionQuery::buildQueryPlan(DB::QueryPlan&) @ 0x000000001cf1c214 in /usr/bin/clickhouse
E           14. /home/ubuntu/clickhouse/src/Interpreters/InterpreterSelectWithUnionQuery.cpp:377: DB::InterpreterSelectWithUnionQuery::execute() @ 0x000000001cf1cd60 in /usr/bin/clickhouse
E           15. /home/ubuntu/clickhouse/src/Interpreters/executeQuery.cpp:1093: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x000000001d40cc4a in /usr/bin/clickhouse
E           16. /home/ubuntu/clickhouse/src/Interpreters/executeQuery.cpp:1238: DB::executeQuery(String const&, std::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) @ 0x000000001d407e44 in /usr/bin/clickhouse
E           17. /home/ubuntu/clickhouse/src/Server/TCPHandler.cpp:500: DB::TCPHandler::runImpl() @ 0x000000001e9946ad in /usr/bin/clickhouse
E           18. /home/ubuntu/clickhouse/src/Server/TCPHandler.cpp:2192: DB::TCPHandler::run() @ 0x000000001e9a7345 in /usr/bin/clickhouse
E           19. /home/ubuntu/clickhouse/base/poco/Net/src/TCPServerConnection.cpp:43: Poco::Net::TCPServerConnection::start() @ 0x0000000024f02459 in /usr/bin/clickhouse
E           20. /home/ubuntu/clickhouse/base/poco/Net/src/TCPServerDispatcher.cpp:115: Poco::Net::TCPServerDispatcher::run() @ 0x0000000024f02cbc in /usr/bin/clickhouse
E           21. /home/ubuntu/clickhouse/base/poco/Foundation/src/ThreadPool.cpp:188: Poco::PooledThread::run() @ 0x00000000250eddd4 in /usr/bin/clickhouse
E           22. /home/ubuntu/clickhouse/base/poco/Foundation/src/Thread.cpp:46: Poco::(anonymous namespace)::RunnableHolder::run() @ 0x00000000250eab7a in /usr/bin/clickhouse
E           23. /home/ubuntu/clickhouse/base/poco/Foundation/src/Thread_POSIX.cpp:335: Poco::ThreadImpl::runnableEntry(void*) @ 0x00000000250e98be in /usr/bin/clickhouse
E           24. ? @ 0x00007f90b06a2ac3 in ?
E           25. ? @ 0x00007f90b0734a40 in ?
E           '.
E           [node2] 2023.11.24 08:50:11.346815 [ 383 ] <Fatal> BaseDaemon: ########################################
E           [node2] 2023.11.24 08:50:11.347040 [ 383 ] <Fatal> BaseDaemon: (version 23.10.1.1, build id: 66D93008E7942432CF4CCFB90BC60C16CF278856, git hash: 77654e906b382ca354a9844e2f1e26b1423d25e0) (from thread 12) (query_id: b35e75e9-c0f1-4eb2-9a12-d4864a699e42) (query: SELECT count() FROM system.remote_data_paths) Received signal Aborted (6)
E           [node2] 2023.11.24 08:50:11.347246 [ 383 ] <Fatal> BaseDaemon:
E           [node2] 2023.11.24 08:50:11.347413 [ 383 ] <Fatal> BaseDaemon: Stack trace: 0x00007f90b06a49fc 0x00007f90b0650476 0x00007f90b06367f3 0x0000000013aebd1e 0x0000000013aef3c9 0x000000001d4075ab 0x000000001d40e281 0x000000001d407e44 0x000000001e9946ad 0x000000001e9a7345 0x0000000024f02459 0x0000000024f02cbc 0x00000000250eddd4 0x00000000250eab7a 0x00000000250e98be 0x00007f90b06a2ac3 0x00007f90b0734a40
E           [node2] 2023.11.24 08:50:11.347665 [ 383 ] <Fatal> BaseDaemon: 4. ? @ 0x00007f90b06a49fc in ?
E           [node2] 2023.11.24 08:50:11.347860 [ 383 ] <Fatal> BaseDaemon: 5. ? @ 0x00007f90b0650476 in ?
E           [node2] 2023.11.24 08:50:11.347996 [ 383 ] <Fatal> BaseDaemon: 6. ? @ 0x00007f90b06367f3 in ?
E           [node2] 2023.11.24 08:50:11.861927 [ 383 ] <Fatal> BaseDaemon: 7. /home/ubuntu/clickhouse/src/Common/Exception.cpp:43: DB::abortOnFailedAssertion(String const&) @ 0x0000000013aebd1e in /usr/bin/clickhouse
E           [node2] 2023.11.24 08:50:11.977998 [ 383 ] <Fatal> BaseDaemon: 8. /home/ubuntu/clickhouse/src/Common/Exception.cpp:447: DB::getCurrentExceptionMessageAndPattern(bool, bool, bool) @ 0x0000000013aef3c9 in /usr/bin/clickhouse
E           [node2] 2023.11.24 08:50:13.255967 [ 383 ] <Fatal> BaseDaemon: 9. /home/ubuntu/clickhouse/src/Interpreters/executeQuery.cpp:586: DB::logExceptionBeforeStart(String const&, std::shared_ptr<DB::Context const>, std::shared_ptr<DB::IAST>, std::shared_ptr<DB::OpenTelemetry::SpanHolder> const&, unsigned long) @ 0x000000001d4075ab in /usr/bin/clickhouse
E           [node2] 2023.11.24 08:50:13.579923 [ 383 ] <Fatal> BaseDaemon: 10. /home/ubuntu/clickhouse/src/Interpreters/executeQuery.cpp:1221: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x000000001d40e281 in /usr/bin/clickhouse
E           [node2] 2023.11.24 08:50:13.926399 [ 383 ] <Fatal> BaseDaemon: 11. /home/ubuntu/clickhouse/src/Interpreters/executeQuery.cpp:1238: DB::executeQuery(String const&, std::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) @ 0x000000001d407e44 in /usr/bin/clickhouse
E           [node2] 2023.11.24 08:50:14.724859 [ 383 ] <Fatal> BaseDaemon: 12. /home/ubuntu/clickhouse/src/Server/TCPHandler.cpp:500: DB::TCPHandler::runImpl() @ 0x000000001e9946ad in /usr/bin/clickhouse
E           [node2] 2023.11.24 08:50:15.133290 [ 383 ] <Fatal> BaseDaemon: 13. /home/ubuntu/clickhouse/src/Server/TCPHandler.cpp:2192: DB::TCPHandler::run() @ 0x000000001e9a7345 in /usr/bin/clickhouse
E           [node2] 2023.11.24 08:50:15.159205 [ 383 ] <Fatal> BaseDaemon: 14. /home/ubuntu/clickhouse/base/poco/Net/src/TCPServerConnection.cpp:43: Poco::Net::TCPServerConnection::start() @ 0x0000000024f02459 in /usr/bin/clickhouse
E           [node2] 2023.11.24 08:50:15.216670 [ 383 ] <Fatal> BaseDaemon: 15. /home/ubuntu/clickhouse/base/poco/Net/src/TCPServerDispatcher.cpp:115: Poco::Net::TCPServerDispatcher::run() @ 0x0000000024f02cbc in /usr/bin/clickhouse
E           [node2] 2023.11.24 08:50:15.284415 [ 383 ] <Fatal> BaseDaemon: 16. /home/ubuntu/clickhouse/base/poco/Foundation/src/ThreadPool.cpp:188: Poco::PooledThread::run() @ 0x00000000250eddd4 in /usr/bin/clickhouse
E           [node2] 2023.11.24 08:50:15.334253 [ 383 ] <Fatal> BaseDaemon: 17. /home/ubuntu/clickhouse/base/poco/Foundation/src/Thread.cpp:46: Poco::(anonymous namespace)::RunnableHolder::run() @ 0x00000000250eab7a in /usr/bin/clickhouse
E           [node2] 2023.11.24 08:50:15.361947 [ 383 ] <Fatal> BaseDaemon: 18. /home/ubuntu/clickhouse/base/poco/Foundation/src/Thread_POSIX.cpp:335: Poco::ThreadImpl::runnableEntry(void*) @ 0x00000000250e98be in /usr/bin/clickhouse
E           [node2] 2023.11.24 08:50:15.362339 [ 383 ] <Fatal> BaseDaemon: 19. ? @ 0x00007f90b06a2ac3 in ?
E           [node2] 2023.11.24 08:50:15.362637 [ 383 ] <Fatal> BaseDaemon: 20. ? @ 0x00007f90b0734a40 in ?
E           [node2] 2023.11.24 08:50:15.362868 [ 383 ] <Fatal> BaseDaemon: Integrity check of the executable skipped because the reference checksum could not be read.
E           [node2] 2023.11.24 08:50:17.499808 [ 383 ] <Fatal> BaseDaemon: This ClickHouse version is not official and should be upgraded to the official build.
E           [node2] 2023.11.24 08:50:17.500562 [ 383 ] <Fatal> BaseDaemon: No settings were changed
E           Error on processing query: Code: 32. DB::Exception: Attempt to read after eof: while receiving packet from 172.16.3.9:9000. (ATTEMPT_TO_READ_AFTER_EOF), Stack trace (when copying this message, always include the lines below):
E
E           0. /home/ubuntu/clickhouse/contrib/llvm-project/libcxx/include/exception:134: std::exception::capture() @ 0x000000000b314402 in /clickhouse
E           1. /home/ubuntu/clickhouse/contrib/llvm-project/libcxx/include/exception:112: std::exception::exception[abi:v15000]() @ 0x000000000b3143cd in /clickhouse
E           2. /home/ubuntu/clickhouse/base/poco/Foundation/src/Exception.cpp:27: Poco::Exception::Exception(String const&, int) @ 0x0000000025070900 in /clickhouse
E           3. /home/ubuntu/clickhouse/src/Common/Exception.cpp:101: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000013aec18e in /clickhouse
E           4. /home/ubuntu/clickhouse/src/Common/Exception.h:85: DB::Exception::Exception(String&&, int, bool) @ 0x000000000b3063aa in /clickhouse
E           5. /home/ubuntu/clickhouse/src/Common/Exception.h:98: DB::Exception::Exception<char const (&) [26]>(int, char const (&) [26]) @ 0x000000000b7527fe in /clickhouse
E           6. /home/ubuntu/clickhouse/src/IO/VarInt.cpp:13: DB::throwReadAfterEOF() @ 0x0000000013bb1611 in /clickhouse
E           7. /home/ubuntu/clickhouse/src/IO/VarInt.h:95: void DB::impl::readVarUInt<true>(unsigned long&, DB::ReadBuffer&) @ 0x0000000013b8ab4a in /clickhouse
E           8. /home/ubuntu/clickhouse/src/IO/VarInt.h:111: DB::readVarUInt(unsigned long&, DB::ReadBuffer&) @ 0x0000000013b8a88c in /clickhouse
E           9. /home/ubuntu/clickhouse/src/Client/Connection.cpp:1035: DB::Connection::receivePacket() @ 0x000000001e81548b in /clickhouse
E           10. /home/ubuntu/clickhouse/src/Client/ClientBase.cpp:1094: DB::ClientBase::receiveAndProcessPacket(std::shared_ptr<DB::IAST>, bool) @ 0x000000001e7ddfab in /clickhouse
E           11. /home/ubuntu/clickhouse/src/Client/ClientBase.cpp:1069: DB::ClientBase::receiveResult(std::shared_ptr<DB::IAST>, int, bool) @ 0x000000001e7ddc2e in /clickhouse
E           12. /home/ubuntu/clickhouse/src/Client/ClientBase.cpp:987: DB::ClientBase::processOrdinaryQuery(String const&, std::shared_ptr<DB::IAST>) @ 0x000000001e7dd198 in /clickhouse
E           13. /home/ubuntu/clickhouse/src/Client/ClientBase.cpp:1844: DB::ClientBase::processParsedSingleQuery(String const&, String const&, std::shared_ptr<DB::IAST>, std::optional<bool>, bool) @ 0x000000001e7dbad3 in /clickhouse
E           14. /home/ubuntu/clickhouse/src/Client/ClientBase.cpp:2108: DB::ClientBase::executeMultiQuery(String const&) @ 0x000000001e7e4c37 in /clickhouse
E           15. /home/ubuntu/clickhouse/src/Client/ClientBase.cpp:2253: DB::ClientBase::processQueryText(String const&) @ 0x000000001e7e6096 in /clickhouse
E           16. /home/ubuntu/clickhouse/src/Client/ClientBase.cpp:2568: DB::ClientBase::runNonInteractive() @ 0x000000001e7e8be0 in /clickhouse
E           17. /home/ubuntu/clickhouse/programs/client/Client.cpp:372: DB::Client::main(std::vector<String, std::allocator<String>> const&) @ 0x0000000013d52d91 in /clickhouse
E           18. /home/ubuntu/clickhouse/base/poco/Util/src/Application.cpp:315: Poco::Util::Application::run() @ 0x0000000024f1d165 in /clickhouse
E           19. /home/ubuntu/clickhouse/programs/client/Client.cpp:1444: mainEntryClickHouseClient(int, char**) @ 0x0000000013d5fe29 in /clickhouse
E           20. /home/ubuntu/clickhouse/programs/main.cpp:491: main @ 0x000000000b2f5665 in /clickhouse
E           21. ? @ 0x00007fd90c86bd90 in ?
E           22. ? @ 0x00007fd90c86be40 in ?
E           23. _start @ 0x000000000b2f4eae in /clickhouse
E            (version 23.10.1.1)
E           (query: SELECT count() FROM system.remote_data_paths)

helpers/client.py:239: QueryRuntimeException
```

PS. Need to check why this error was not seen in CI

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Closes https://github.com/ClickHouse/ClickHouse/issues/57188.

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 4d64cd5d116eb1df36b02192433ad340266d3647 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57322/4d64cd5d116eb1df36b02192433ad340266d3647/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7021692143">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57322/4d64cd5d116eb1df36b02192433ad340266d3647/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Disks/ObjectStorages/Web/WebObjectStorage.cpp [hunk 1]
```diff
@@ -124,7 +124,19 @@ bool WebObjectStorage::exists(const StoredObject & object) const
...
 {
     LOG_TRACE(&Poco::Logger::get("DiskWeb"), "Checking existence of path: {}", path);
+    return tryGetFileInfo(path) != std::nullopt;
+}
+
+WebObjectStorage::FileData WebObjectStorage::getFileInfo(const String & path) const
+{
+    auto file_info = tryGetFileInfo(path);
+    if (!file_info)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "No such file: {}", path);
```
Comment (devcrafter) [inline src/Disks/ObjectStorages/Web/WebObjectStorage.cpp:134]: Why is it a logical error? Do we have any guarantees that the file should be there, or what are our assumptions?

Comment (kssenii) [reply src/Disks/ObjectStorages/Web/WebObjectStorage.cpp]: yes, better to use `FILE_DOESNT_EXIST`.
In general it should be checked for `exists` method or be sure that file exists as it was a result of `list` or `iterateDirectory` methods.

File: src/Disks/ObjectStorages/Web/WebObjectStorage.cpp [hunk 2]
```diff
@@ -176,10 +189,9 @@ bool WebObjectStorage::exists(const std::string & path) const
         unique_lock.unlock();
         shared_lock.lock();
 
-        return true;
+        return FileData{ .type = FileType::Directory };
```
Comment (devcrafter) [inline src/Disks/ObjectStorages/Web/WebObjectStorage.cpp:192]: The comment above need to be fixed, it refers to returning `bool` type

**Review by devcrafter — APPROVED** [review]
LGTM

**Pull request merged**